### ISSUE TITLE
Outfeed with tuple shape on CPU never finishes

### ISSUE THREAD

**Issue by seanmor5** [google/jax#7740]
Hello,

I am almost positive this is a bug with XLA's CPU Xfeed Manager. Here is a repro from `infeed_test.py` changed to outfeed a tuple:

```
@jax.jit
def f(x):
  token = lax.create_token(x)
  y, token = lax.infeed(
      token, shape=jax.ShapedArray((3, 4), jnp.float32))
  token = lax.outfeed(token, (y + np.float32(1), y,))
  return x - 1

x = np.float32(7.5)
y = np.random.randn(3, 4).astype(np.float32)
execution = threading.Thread(target=lambda: f(x))
execution.start()
device = jax.local_devices()[0]
device.transfer_to_infeed(y)
out_shape = xla_client.shape_from_pyval(y)
tup_shape = xla_client.Shape.tuple_shape([out_shape, out_shape])
out = device.transfer_from_outfeed(tup_shape.with_major_to_minor_layout_if_absent())
execution.join()
```

This will run forever on CPU. I have not tested on GPU or TPU yet. I believe the problem is the fact that the CPU Xfeed Manager calls `WaitForNotification` here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/cpu/cpu_xfeed.cc#L150

but `Done`: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/cpu/cpu_xfeed.cc#L70

which is supposed to send the notification is never called.

**Comment by jekbradbury** [comment]
That sounds plausible! I'm tentatively assigning someone from the runtime side to take a look.

**Comment by seanmor5** [comment]
Strangely enough, I just re-ran the Colab Notebook from scratch and it worked - not really sure what changed...

**Comment by zhangqiaorjc** [comment]
I tried this locally and it's fine.

```
➜  test python issue7740.py
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
finished
```

Notice that infeed and outfeed are stateful operations, so if your colab cell execution got interrupted due to say ImportError, and you simply re-run the same cell a second time, there's no guarantee what will happen.

I'm going to close this for now. Please re-open if it reoccurs.

**Comment by zhangqiaorjc** [comment]
a note for future: this actually repro'ed once on colab and colab only, it doesn't happen otherwise and disappears after colab restart; since it rarely repros, it might take a while to find root cause

what i noticed is that the colab log shows

```
I0907 14:39:08.500906 520884 main/520884 cpu_xfeed.cc:178 Transferring literal to infeed with shape: f32[3,4]
I0907 14:39:08.504105 520884 main/520884 xfeed_manager.cc:44 Enqueueing infeed buffer (of 1 buffers) with length: 48
I0907 14:39:08.506614 520884 main/520884 cpu_xfeed.cc:132 Enqueueing outfeed buffer (for the device to populate) of length 48B
I0907 14:39:08.506631 520884 main/520884 cpu_xfeed.cc:132 Enqueueing outfeed buffer (for the device to populate) of length 48B
I0907 14:39:08.506634 520884 main/520884 xfeed_manager.cc:44 Enqueueing outfeed buffer (of 2 buffers) with length: 48
I0907 14:39:08.506641 520884 main/520884 xfeed_manager.cc:44 Enqueueing outfeed buffer (of 2 buffers) with length: 48
I0907 14:39:08.506644 520884 main/520884 cpu_xfeed.cc:147 Waiting for buffer to be notified as populated.
```

And while the internal log otherwise is interleaved with lots of XLA/CPU compiler log output. I wonder if python thread gets stuck on outfeed dequeue wait before that same thread is asked to compile the program so it deadlocks. It's possible that there's some rare race condition that only happens when there's a single core available.

@hawkinsp

**Comment by seanmor5** [comment]
@zhangqiaorjc We are experiencing a race condition with XLA running on the GPU in the Erlang VM with a program that is very similar to this. We launch the compilation/execution in a separate VM process, and then continue with `infeed`/`outfeed` ops in the main process. Sometimes it finishes; however, other times the operations deadlock (presumably because the computation is stuck waiting for an infeed from the main process and the main process is waiting for an outfeed from the computation). I'm not up to speed on the internals of Python processes versus threads and how that relates to the Erlang VM's pre-emptive scheduler; however, running the above program using Python multiprocessing instead of threads causes it to deadlock for me 100% of the time in Google Colab:

```python
from jaxlib import xla_client
from jax import lax
import jax
import time
import numpy as np
import jax.numpy as jnp
import multiprocessing as mp
from multiprocessing import Process

mp.set_start_method('forkserver') # Tried with 'spawn' as well

# Indirection for pickling?
def f_p(x_p):

  @jax.jit
  def f(x):
    token = lax.create_token(x)
    y, token = lax.infeed(
        token, shape=jax.ShapedArray((3, 4), jnp.float32))
    token = lax.outfeed(token, (y + np.float32(1), y,))
    return x - 1

  return f(x_p)

x = np.float32(7.5)
y = np.random.randn(3, 4).astype(np.float32)
execution = Process(target=f_p, args=(x,))
execution.start()
device = jax.local_devices()[0]
device.transfer_to_infeed(y)
out_shape = xla_client.shape_from_pyval(y)
tup_shape = xla_client.Shape.tuple_shape([out_shape, out_shape])
out = device.transfer_from_outfeed(tup_shape.with_major_to_minor_layout_if_absent())
execution.join()
```

Commenting out the outfeed operation and transfer_from_outfeed and the computation completes as normal

**Comment by hawkinsp** [comment]
@seanmor5 JAX has no support for multiprocess access to infeed/outfeed queues on any platform.
